software/earl/authenticator.go

```diff
@@ -120,21 +135,26 @@ func (a *FileBasedAuthenticator) IterateUsers(callback func(user User)) {
 // Check if access for a given code is granted to a given Target
 func (a *FileBasedAuthenticator) AuthUser(code string, target Target) (AuthResult, string) {
 	if !hasMinimalCodeRequirements(code) {
+		authCounter.WithLabelValues(target.String(), "too short code").Inc()
 		return AuthFail, "Auth failed: too short code."
 	}
 	user := a.findUserSynchronized(code, nil)
 	if user == nil {
+		authCounter.WithLabelValues(target.String(), "no user for code").Inc()
 		return AuthFail, "No user for code"
 	}
 	// In case of Hiatus users, be a bit more specific with logging: this
 	// might be someone stolen a token of some person on leave or attempt
 	// of a blocked user to get access.
 	if user.UserLevel == LevelHiatus {
+		authCounter.WithLabelValues(target.String(), "user on hiatus").Inc()
 		return AuthFail, fmt.Sprintf("User on hiatus '%s <%s>'", user.Name, user.ContactInfo)
 	}
 	if !user.InValidityPeriod(a.clock.Now()) {
+		authCounter.WithLabelValues(target.String(), "expired").Inc()
 		return AuthExpired, "Code not valid yet/expired"
 	}
+	authCounter.WithLabelValues(target.String(), "OK").Inc()
```

Comment:
At this point, it is actually not yet OK, as we await the result from `userHasAccess()`, which can tell if we are within the right time of day.

instead of having to put instrumentation in here, which requires to make sure to catch every return, how about we rename this `authUserInternal()` and then have `AuthUser()` itself call that internal function, capture the result and call the authCounter() with the target, AuthResult and/or string. That way, the whole instrumentation is only a few lines of code which won't miss a return code if needed.

Is it possible with Prometheus to log 'enums' (in quotes, because Go does not have enums) instead of the strings ? I think it would be easier to use with automatic tools than the string-matching. So if we could map the numbers to short strings such as `AuthFail` that would be useful.

Reply:
The problem here is that the code has only one `AuthFailed`, but there are multiple types of failures. We need to capture the individual types of failures.

We could use a defer, and name the return string, and pass that for the label. But one of the returned strings includes user info, which is not what we want as we're trying to capture the various failure modes.

Prometheus has an enum option, but this is not what we want here, we want counters. Either way they end up in labels. This is standard for metrics, and makes no difference. Prometheus and Grafana are the tools we use with this.

Reply:
So what if `authUserInternal()` would return a tuple with three results `(AuthResult, string, string)`, which would allow to have the long string and short monitoring string returned, e.g
```
AuthExpired, "Code not valid yet/expired", "expired"
```
```
return AuthFail, fmt.Sprintf("User on hiatus '%s <%s>'", user.Name, user.ContactInfo), "hiatus"
```
(also userHasAccess() needs to return the triple)

Then we can use that for the monitoring and the regular auth, something like:
```
  func (a *FileBasedAuthenticator) AuthUser(code string, target Target) (AuthResult, string) {
     result, auth_msg, monitor_label := a.authUserInternal(code, target)
     authCounter.WithLabelValues(target.String(), monitor_label).Inc()
     return result, auth_msg
  }
```
I would make the monitor labels simple strings without spaces to save processing headaches later on when processing with any kind of script that has whitespace separation. So not "too short code" but "code_too_short". Saves potential quoting hell; not sure if this actually a problem in practice, but having identifier-like labels usually pays off.